scipy/spatial/tests/test_distance.py

```diff
@@ -678,301 +946,417 @@ def test_pdist_mahalanobis(self):
+        m2 = wrussellrao(np.array([1, 0, 1], dtype=bool),
+                         np.array([1, 1, 0], dtype=bool))
         if verbose > 2:
             print(m)
         assert_allclose(m, 2/3, rtol=0, atol=1e-10)
         assert_allclose(m2, 2/3, rtol=0, atol=1e-10)
 
+    def test_pdist_russellrao_match(self):
+        D = eo['random-bool-data']
+        if verbose > 2:
+            print(D.shape, D.dtype)
+        eps = 1e-10
+        y1 = wpdist(D, "russellrao")
+        y2 = wpdist(D, "test_russellrao")
+        y3 = wpdist(np.bool_(D), "test_russellrao")
+        if verbose > 2:
+            print(np.abs(y1-y2).max())
+            print(np.abs(y2-y3).max())
+        _assert_within_tol(y1, y2, eps)
+        _assert_within_tol(y2, y3, eps)
+
+    def test_pdist_sokalmichener_match(self):
+        D = eo['random-bool-data']
+        if verbose > 2:
+            print(D.shape, D.dtype)
+        eps = 1e-10
+        y1 = wpdist(D, "sokalmichener")
+        y2 = wpdist(D, "test_sokalmichener")
+        y3 = wpdist(np.bool_(D), "test_sokalmichener")
+        if verbose > 2:
+            print(np.abs(y1-y2).max())
+            print(np.abs(y2-y3).max())
+        _assert_within_tol(y1, y2, eps)
+        _assert_within_tol(y2, y3, eps)
+
+    def test_pdist_kulsinski_match(self):
+        D = eo['random-bool-data']
+        if verbose > 2:
+            print(D.shape, D.dtype)
+        eps = 1e-10
+        y1 = wpdist(D, "kulsinski")
+        y2 = wpdist(D, "test_kulsinski")
+        y3 = wpdist(np.bool_(D), "test_kulsinski")
+        _assert_within_tol(y1, y2, eps, verbose > 2)
+        _assert_within_tol(y2, y3, eps)
+
     def test_pdist_canberra_match(self):
         D = eo['iris']
         if verbose > 2:
             print(D.shape, D.dtype)
         eps = 1e-10
-        y1 = pdist(D, "canberra")
-        y2 = pdist(D, "test_canberra")
+        y1 = wpdist_no_const(D, "canberra")
+        y2 = wpdist_no_const(D, "test_canberra")
         _assert_within_tol(y1, y2, eps, verbose > 2)
 
     def test_pdist_canberra_ticket_711(self):
         # Test pdist(X, 'canberra') to see if Canberra gives the right result
         # as reported on gh-1238.
         eps = 1e-8
-        pdist_y = pdist(([3.3], [3.4]), "canberra")
+        pdist_y = wpdist_no_const(([3.3], [3.4]), "canberra")
         right_y = 0.01492537
         _assert_within_tol(pdist_y, right_y, eps, verbose > 2)
 
-    def test_pdist_custom_notdouble(self):
-        # tests that when using a custom metric the data type is not altered
-        class myclass(object):
-            pass
-
-        def dummy_metric(x, y):
-            if not isinstance(x[0], myclass) or not isinstance(y[0], myclass):
-                raise ValueError("Type has been changed")
-            return 1.123
-        data = np.array([[myclass()], [myclass()]], dtype=object)
-        pdist_y = pdist(data, metric=dummy_metric)
-        right_y = 1.123
-        assert_equal(pdist_y, right_y, verbose=verbose > 2)
-
```

Comment:
I added a comment about this above:
When I was working on tests, I noticed #1403 -- I've changed behavior so that now completely dissimilar vectors get inf distance in yule, dice, and sokalsneath. Previously, sokalsneath would raise a ValueError and yule and dice would raise ZeroDivisionError's.